Document package types (#2246)
jpmorganchase/modular: docs/concepts/package-types.md

---
title: Types of packages
parent: Concepts
---

> 

Author: Done.

> | Type       | `modular start` | `modular build` | `modular test` | `modular lint` | Custom index / assets      | Bundled                                      | Entrypoint                                                                                                                                                                                                                                                                                       |

Author: Done

| ---------- | --------------- | --------------- | -------------- | -------------- | -------------------------- | -------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |

> | `app`      | Yes             | Yes             | Yes            | Yes            | Yes                        | Yes                                          | The compiled bundle will use `src/index.tsx` as an entrypoint and include it in the output `index.html`                                                                                                                                                                                          |

Reviewer: Include in the _outputted_ index.html or include it in the _output of_ index.html?

> | `esm-view` | Yes             | Yes             | Yes            | Yes            | No                         | Yes + rewriting external dependencies to CDN | The compiled bundle will use `src/index.tsx` as an entrypoint and [link it](../esm-views/output-package-manifest.md) in the output `package.json`; the entrypoint needs to default export a React component for `modular start`/ [standalone `index.html`](../esm-views/how-to-build.md) to work |

Reviewer: The expression "default export a React component" might not be obvious to everyone. Maybe it can be more explicit?

e.g.
> the entrypoint needs to export a default (non-named) React component

> | `esm-view` | Yes             | Yes             | Yes            | Yes            | No                         | Yes + rewriting external dependencies to CDN | The compiled bundle will use `src/index.tsx` as an entrypoint and [link it](../esm-views/output-package-manifest.md) in the output `package.json`; the entrypoint needs to default export a React component for `modular start`/ [standalone `index.html`](../esm-views/how-to-build.md) to work |

Author: This *is* outdated: was rewritten already as "the entry-point needs to export a default (non-named) React component"

> | `esm-view` | Yes             | Yes             | Yes            | Yes            | No                         | Yes + rewriting external dependencies to CDN | The compiled bundle will use `src/index.tsx` as an entrypoint and [link it](../esm-views/output-package-manifest.md) in the output `package.json`; the entrypoint needs to default export a React component for `modular start`/ [standalone `index.html`](../esm-views/how-to-build.md) to work |

Author: Good point, but I prefer "Yes. Will rewrite external dependencies to point to a CDN" because by default Modular rewrites them and bundling external dependencies has unintuitive ramifications (described in the "known limitations" section) and is an escape hatch that needs to be used only in rare situations.

> | `package`  | No              | Yes             | Yes            | Yes            | N/A                        | No                                           | Whatever specified in `main` field of `package.json` is used as entrypoint                                                                                                                                                                                                                       |

Reviewer: ```suggestion
| `package`  | No              | Yes             | Yes            | Yes            | N/A                        | No                                           | Whatever is specified in `main` field of `package.json` is used as the entrypoint                                                                                                                                                                                                                       |
```

> | `view`     | Yes             | Yes             | Yes            | Yes            | N/A                        | No                                           | Whatever specified in `main` field of `package.json` is used as entrypoint; needs to default export React component for `modular start` to work                                                                                                                                                  |

Reviewer: ```suggestion
| `view`     | Yes             | Yes             | Yes            | Yes            | N/A                        | No                                           | Whatever is specified in `main` field of `package.json` is used as the entrypoint; needs to default export React component for `modular start` to work                                                                                                                                                  |
```

Also: same feedback as above regarding the expression of "default export" as a verb

> | `view`     | Yes             | Yes             | Yes            | Yes            | N/A                        | No                                           | Whatever specified in `main` field of `package.json` is used as entrypoint; needs to default export React component for `modular start` to work                                                                                                                                                  |

Author: Done.

| `source`   | No              | No              | Yes            | Yes            | N/A                        | N/A - never built                            | N/A - don't need the concept of entrypoint since this package is never built                                                                                                                                                                                                                     |
| `template` | No              | No              | Yes            | Yes            | Depends on the target type | N/A - never built                            | N/A - don't need the concept of entrypoint since this is a special type of package used by `modular add` which is never built directly                                                                                                                                                           |

## App

> Modular `app`s are Typescript and React based Single Page Applications (SPA)

Reviewer: Nitpicking... 😄 

```suggestion
Modular `app`s are TypeScript and React-based Single Page Applications (SPAs)
```

> Modular `app`s are Typescript and React based Single Page Applications (SPA)

Author: And this is outdated too

built with [Webpack v5](https://webpack.js.org/) (by default) or
[esbuild](https://esbuild.github.io/) (by turning on `useModularEsbuild` in
[the configuration](../configuration.md)).

> Apps are generated by `modular add` using the

Reviewer: I think this could be 2 sentences to read a bit easier.

The 2nd sentence could start like this:

> ... templates and need at least a ...

becomes

> ... templates. They need at least ...

Author: Split the sentence and moved the template part to the bottom, like all the other types.

> [`modular-template-app`](https://github.com/jpmorganchase/modular/tree/main/packages/modular-template-app)

Author: @benpryke they **are** mentioned for other types. Every type has, at the end, a paragraph describing which template creates it. Again, you're referring to an outdated copy of the document:

> Apps are generated by `modular add` using the
[`modular-template-app`](https://github.com/jpmorganchase/modular/tree/main/packages/modular-template-app)
[template](../concepts/templates.md)

> ESM Views are generated by `modular add` using the
[`modular-template-esm-view`](https://github.com/jpmorganchase/modular/tree/main/packages/modular-template-esm-view)
[template](../concepts/templates.md)

> Packages are generated by `modular add` using the
[`modular-template-package`](https://github.com/jpmorganchase/modular/tree/main/packages/modular-template-package)
[template](../concepts/templates.md)

> Views are generated by `modular add` using the
[`modular-template-view`](https://github.com/jpmorganchase/modular/tree/main/packages/modular-template-view)
[template](../concepts/templates.md)

> Sources are generated by `modular add` using the
[`modular-template-source`](https://github.com/jpmorganchase/modular/tree/main/packages/modular-template-source)
[template](../concepts/templates.md)

> Templates are created from scratch;
to find out how to set up one, please follow
[this guide](../how-to/create-template.md)

[template](../concepts/templates.md) and need at least a `public/index.html`
file and an entrypoint located in `src/index.tsx`, that typically `render`s a
React component to the DOM, to work.

> Apps support pretty much all the functionalities from

Reviewer: Just a thought on this.

Not sure if this is covered by other docs but it might be good to be a bit more transparent about how modular-scripts relates to CRA. Also, this isn't specific to `apps` but obviously underpins them (and the other types), so not sure where it would go.

I'm thinking we might want to say something like:

> Modular is based on Create React App v5 and maintains a high level of compatibility, but is an independent project that makes it's own choices (and has it's own features) that occasional diverge from Create React App ... _details_

I say this because 
- It's a good starting place if the user is familiar with CRA, and they can refer to that project and their existing knowledge
- We make it clear that it's not literally CRA, and they should refer to these docs

[Create React App v5](https://create-react-app.dev/docs/custom-templates), with
some important differences:

> 

Reviewer: Maybe not something we need to do right now, but it would be good to have the _why_ for each point. Developers might be wondering why we don't have eject, why babel plugins are not supported, etc

Author: Added some "whys"

- [ejecting](https://create-react-app.dev/docs/available-scripts/#npm-run-eject)
  is not supported
- [template support](../concepts/templates.md) is integrated in the
  [`modular add`](../commands/add.md) command. Modular templates are not
  guaranteed to be compatible with CRA templates.
- Source files are loaded with the more performant
  [`esbuild-loader`](https://github.com/privatenumber/esbuild-loader) in the
  Webpack configuration, and Babel plugins are not supported.
- [Local proxies](https://create-react-app.dev/docs/proxying-api-requests-in-development/)
  are supported, but [esbuild mode](../configuration.md) doesn't support the
  `package.json` `proxy` field. The more flexible
  [manual proxy configration](https://create-react-app.dev/docs/proxying-api-requests-in-development/#configuring-the-proxy-manually)
  is supported in both Webpack and esbuild mode.

> Apps can be built using [`modular build`](../commands/build.md): the resulting

Author: Done.

output is an optimized site that can be served statically. All code (files in
`src` plus external dependencies required in the code) are bundled in a single
blob of code that can be split in different files.

> Apps can be started using [`modular start`](../commands/start.md): a developer

Author: Done

server runs on port 3000, serving the app with an additional runtime layer that
provides developer experience functionalities like hot reloading and on-screen
error overlay.

## ESM View

> Modular `esm-view`s are similar to `app`s with some essential differences: they

Author: Done, see below.

are compiled as [ES
Modules(https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules)]
and all their external `import`s are rewritten to point to an

> [ESM CDN](../esm-views/esm-cdn.md). Like apps, they are be built with the same

Author: Done.

Webpack or esbuild configuration and support the same functionalities. Unlike

> apps, they don't allow the user to specify a custom `index.html`, they don't

Author: Collated all differences in a ordered list.

include a `public` folder in the build and they expect their entrypoint
(`src/index.tsx`) to not `render` to the DOM, but to export something, typically
a React Component.

> Esm Views are Modular's type of choice to implement the

Author: Done.

> [micro-frontend](../concepts/microfrontends.md) pattern. They typically expect

Author: As discussed: will create a section for package types but not nest esm-views

> [micro-frontend](../concepts/microfrontends.md) pattern. They typically expect

Author: Done.

their entrypoint exports to be
[dynamically `import`ed at runtime](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import)
by some other micro-frontend, but can be also
[served standalone](../esm-views/how-to-build.md) exactly as an `app`, provided
that they default export a React component that can be rendered to the DOM.

> Since the micro-frontend pattern allows different teams to build and serve their

Reviewer 2: This paragraph is very wordy and I fear it will get lost. Is there a better way to provide this information? E.g. in a "How to use a CDN for external dependencies" guide or something?

Author: I fear we don't have such a resource. Do you wanna try to simplify this? I would like to convey the message that we have esm views because they can be built in isolation and composed at run-time.

applications independently and compose them at runtime (parallelizing otherwise
expensive build operations), third-party dependencies need to be de-duplicated
across micro-frontends; for this reason, external dependencies are not bundled,
but they are rewritten to a [configurable ESM CDN](../esm-views/esm-cdn.md) that
serves them as ES modules on the fly.

> Esm Views can be built using [`modular build`](../commands/build.md): the

Reviewer 2: Following my comment about separate pages for each package type, I think each page should have headings for each of the [appropriate?] Modular commands (always in the same order), that explain the detail here and include a code snippet for e.g. `modular build my-view`, `modular add --type esm-view` etc. Don't worry about the duplication between pages; this is on purpose: some users will only view one page, and those who view several will quickly learn about the common interface that Modular exposes.

Optionally, you could link the answers in the table above to each specific section on the page to provide additional detail, if desired, instead of using longer sentences in the table.

Author: - The problem with `add` is that we have `--unstable-type`, derived from CRA, that we don't really want to show on the docs. The preferred mode of adding a page is the interactive prompt.
- The problem with `build` is that the command will always be the same.

But I like the idea of having structured information: I can do that for entry-points and to explain what happens when things are built or started, for example.

resulting output is an optimized site that can be either served statically
(using a synthetically-generated `index.html` file and a `_trampoline.js`

> loader), or dynamically imported by another ESM view. The js and CSS

Reviewer: Grammatical nitpick

```suggestion
loader), or dynamically imported by another ESM view. The JS and CSS
```

entrypoints, in the second case, are linked in the
[generated manifest file](../esm-views/output-package-manifest.md), that the
loading code can fetch and examine before loading the `esm-view`. When building
an ESM View, all "local" code (files in `src`) is bundled in a single blob that
can be split into different files. All external dependencies are rewritten to an
external CDN by default and don't get bundled, although this behavior is
[configurable](../configuration.md).

ESM Views are generated by `modular add` using the
[`modular-template-esm-view`](https://github.com/jpmorganchase/modular/tree/main/packages/modular-template-esm-view)
[template](../concepts/templates.md)

## Package

Modular `package`s are generic, publishable libraries with a single entrypoint
file. They are built with [Rollup.js](https://rollupjs.org/guide/en/) and they
are not bundled in a single blob; files required directly or indirectly from the

> entry point are separately transpiled as CJS Modules and ES Modules (althought

Reviewer: Consistency thing, plus a typo

```suggestion
entry point are separately transpiled as CommonJS Modules and ES Modules (although
```

> it's not a 1:1 correspondance: the compilation process typically creates

Reviewer: ```suggestion
it's not a 1:1 correspondence: the compilation process typically creates
```

Reviewer: I _think_ it's supposed to be spelled that way. Not sure if US/UK thing

Author: "Correspondance" is the French word, so you're right :)

building a Modular `package`:

```json
{
  "name": "<your-package-name>",
  "private": false,
  "modular": {
    "type": "package"
  },
  "main": "dist-cjs/index.js",
  "version": "1.0.0",
  "module": "dist-es/index.js",
  "typings": "dist-types/index.d.ts",
  "dependencies": {},
  "files": ["dist-cjs", "dist-es", "dist-types", "README.md"]
}
```

### When to use the "package" type

Modular `package`s are meant to provide re-usable functionality that can be
published to third-party registries (like
[the npm registry](https://www.npmjs.com/)) out-of-the-box; to underline this,
Modular will, for example, refuse to work with packages that have the
[`private`](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#private)
field set. You should use packages _only when there is a need to consume their
build output_; if you just need to share some common source code inside your
monorepo but you don't need to publish it externally, you should use the
`source` type instead.

## View

Modular `view`s are Modular `package`s that, by convention, export a default
React component. They are [created](../commands/add.md) with a default

> entrypoint is `src/index.tsx` in their manifest and they are built and tested

Reviewer: I assume by manifest you actually mean `package.json` btw?

exactly like `package` types. The only difference is that, since `view`s should
always export a React component, they can be [`start`ed](../commands/start.md)
to spawn a local developer server and render their default export to the DOM.

Packages are generated by `modular add` using the
[`modular-template-view`](https://github.com/jpmorganchase/modular/tree/main/packages/modular-template-view)
[template](../concepts/templates.md)

## Source

> Modular `source` types are packages that contain only source code in their `src`

Reviewer 2: Don't `package`s also only contain source code in their `src` directory? I think the defining feature of the `source` type is: Modular `source` packages contain common source code that can be imported from other packages in the monorepo but cannot be built themselves.

Author: The subject of the adversative seems to refer to the "that" in the relative clause and sound a bit confusing. What do you think of:

> Modular `source` packages contain common source code that can be imported from
other packages in the monorepo. Source packages are not meant to be built
themselves; they are meant to be imported directly from other packages in the
monorepo.

directory; they are meant to be imported directly from other packages in the
monorepository and can be [`test`ed](../commands/test.md), but not
[`built`](../commands/build.md) or [`start`ed](../commands/start.md). `source`s
are useful to factor out some common code that's used in a monorepository and
doesn't need to be built separately or published to a registry. `source`s can
also be used to temporarily work around
[circular dependency errors](../concepts/circular-dependencies.md), although
their use in this sense is limited and it's always prefereable to factor out the
common parts of a circular dependency in an additional dependency to break the
cycle.

Packages are generated by `modular add` using the
[`modular-template-source`](https://github.com/jpmorganchase/modular/tree/main/packages/modular-template-source)
[template](../concepts/templates.md)

## Template

The Modular `template` type is a special type that is never directly created by
[`modular add`](../commands/add.md), but it's used by it as a blueprint to
create other packages of various types. An in-depth explanation of templates can

> be found [here](../concepts/templates.md). Templates are created from scratch;

Reviewer 2: If we use a `docs/package-types` structure, then we can merge `docs/package-types/template.md` (this section of this current doc) with the existing `docs/concepts/templates.md`.

Author: Good idea, done.

| ---------- | --------------- | --------------- | -------------- | -------------- | -------------------------- | -------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `app`      | Yes             | Yes             | Yes            | Yes            | Yes                        | Yes                                          | The compiled bundle will use `src/index.tsx` as an entrypoint and include it in the generated `index.html`                                                                                                                                                                                                    |

> | `esm-view` | Yes             | Yes             | Yes            | Yes            | No                         | Yes + rewriting external dependencies to CDN | The compiled bundle will use `src/index.tsx` as an entrypoint and [link it](../esm-views/output-package-manifest.md) in the output `package.json`; the entrypoint needs to export a default (non-named) React componentt for `modular start`/ [standalone `index.html`](../esm-views/how-to-build.md) to work |

Reviewer: New typo

```suggestion
| `esm-view` | Yes             | Yes             | Yes            | Yes            | No                         | Yes + rewriting external dependencies to CDN | The compiled bundle will use `src/index.tsx` as an entrypoint and [link it](../esm-views/output-package-manifest.md) in the output `package.json`; the entrypoint needs to export a default (non-named) React component for `modular start`/ [standalone `index.html`](../esm-views/how-to-build.md) to work |
```

some important differences:

- [ejecting](https://create-react-app.dev/docs/available-scripts/#npm-run-eject)
  is not supported. As an opinionated tool, Modular tries to offer an uniform
  way of building applications, although feedback on our build configuration is
  welcome!
- [template support](../concepts/templates.md) is integrated in the
  [`modular add`](../commands/add.md) command. Modular templates are not
  guaranteed to be compatible with CRA templates.
- Source files are loaded with the more performant
  [`esbuild-loader`](https://github.com/privatenumber/esbuild-loader) in the

>   Webpack configuration. for this reason, Babel plugins are not supported.

Reviewer: ```suggestion
  Webpack configuration. For this reason, Babel plugins are not supported.
```
